addons/score/spoke/calculate.go

```diff
@@ -0,0 +1,220 @@
+package spoke
+
+import (
+	"os"
+	"strconv"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	corev1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+const (
+	maxScore              = float64(100)
+	minScore              = float64(-100)
+	defaultMaxCPUCount    = "100"
+	defaultMinCPUCount    = "0"
+	defaultMaxMemoryCount = "1048576"
+	defaultMinMemoryCount = "0"
+)
+
+type Score struct {
+	nodeLister        corev1lister.NodeLister
+	useRequested      bool
+	enablePodOverhead bool
+	podListener       corev1lister.PodLister
+}
+
+func NewScore(nodeInformer corev1informers.NodeInformer, podInformer corev1informers.PodInformer) *Score {
+	return &Score{
+		nodeLister:        nodeInformer.Lister(),
+		podListener:       podInformer.Lister(),
+		enablePodOverhead: true,
+		useRequested:      true,
+	}
+}
+
+func (s *Score) calculateScore() (cpuScore int64, memScore int64, err error) {
+	cpuAlloc, err := s.calculateClusterAllocateable(clusterv1.ResourceCPU)
+	if err != nil {
+		return 0, 0, err
+	}
+	memAlloc, err := s.calculateClusterAllocateable(clusterv1.ResourceMemory)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	cpuUsage, err := s.calculatePodResourceRequest(v1.ResourceCPU)
+	if err != nil {
+		return 0, 0, err
+	}
+	memUsage, err := s.calculatePodResourceRequest(v1.ResourceMemory)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return s.normalizeScore(cpuAlloc, cpuUsage, memAlloc, memUsage)
+}
+
+func (s *Score) normalizeScore(cpuAlloc, cpuUsage, memAlloc, memUsage float64) (cpuScore int64, memScore int64, err error) {
+	klog.Infof("cpuAlloc = %v, cpuUsage = %v, memAlloc = %v, memUsage = %v", cpuAlloc, cpuUsage, int64(memAlloc), int64(memUsage))
+	availableCpu := cpuAlloc - cpuUsage
+
+	maxCPUCountNum, minCPUCountNum, maxMemoryCountNum, minMemoryCountNum, err := s.getMaxMinValues()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if availableCpu > maxCPUCountNum {
+		cpuScore = int64(maxScore)
+	} else if availableCpu <= minCPUCountNum {
+		cpuScore = int64(minScore)
+	} else {
+		cpuScore = int64(200*availableCpu/maxCPUCountNum - 100)
+	}
+
+	availableMem := (memAlloc - memUsage) / (1024 * 1024) // MB
+	if availableMem > maxMemoryCountNum {
+		memScore = int64(maxScore)
+	} else if availableMem <= minMemoryCountNum {
+		memScore = int64(minScore)
+	} else {
+		memScore = int64(200*availableMem/maxMemoryCountNum - 100)
+	}
+
+	klog.Infof("cpuScore = %v, memScore = %v", cpuScore, memScore)
+	return cpuScore, memScore, nil
+}
+
+func (s *Score) calculateClusterAllocateable(resourceName clusterv1.ResourceName) (float64, error) {
+	nodes, err := s.nodeLister.List(labels.Everything())
+	if err != nil {
+		return 0, err
+	}
+
+	allocatableList := make(map[clusterv1.ResourceName]resource.Quantity)
+	for _, node := range nodes {
+		if node.Spec.Unschedulable {
+			continue
```

Comment:
Its bad practice to use  continue ... change the if structure to `if  !node.spec.Unschedulable`